src/realm/object-store/c_api/app.cpp

```diff
@@ -826,4 +828,230 @@ RLM_API realm_app_t* realm_user_get_app(const realm_user_t* user) noexcept
     return nullptr;
 }
 
+template <typename T>
+inline T convert_to_bson(realm_string_t* doc)
+{
+    if (doc == nullptr || doc->data == nullptr) {
+        return T();
+    }
+    return static_cast<T>(bson::parse({doc->data, doc->size}));
+}
+
+static MongoCollection::FindOptions
+to_mongodb_collection_find_options(realm_mongodb_find_and_modify_options_t* options)
+{
+    MongoCollection::FindOptions mongodb_options;
+
+    if (options->projection_bson != nullptr) {
+        mongodb_options.projection_bson = convert_to_bson<bson::BsonDocument>(options->projection_bson);
+    }
+
+    if (options->sort_bson != nullptr) {
+        mongodb_options.sort_bson = convert_to_bson<bson::BsonDocument>(options->sort_bson);
+    }
+
+    if (options->limit != 0) {
+        mongodb_options.limit = options->limit;
+    }
+
+    return mongodb_options;
+}
+
+static MongoCollection::FindOneAndModifyOptions
+to_mongodb_collection_find_and_modify_options(realm_mongodb_find_and_modify_options_t* options)
+{
+    MongoCollection::FindOneAndModifyOptions mongodb_options;
+
+    if (options->projection_bson != nullptr) {
+        mongodb_options.projection_bson = convert_to_bson<bson::BsonDocument>(options->projection_bson);
+    }
+
+    if (options->sort_bson != nullptr) {
+        mongodb_options.sort_bson = convert_to_bson<bson::BsonDocument>(options->sort_bson);
+    }
+
+    mongodb_options.upsert = options->upsert;
+    mongodb_options.return_new_document = options->return_new_document;
+
+    return mongodb_options;
+}
+
+static void handle_mongodb_collection_result(util::Optional<bson::Bson> bson, util::Optional<AppError> app_error,
+                                             realm_userdata_t data, realm_free_userdata_func_t delete_data,
+                                             realm_mongodb_callback_t callback)
+{
+    if (app_error) {
+        auto error = to_capi(*app_error);
+        callback(data, nullptr, &error);
+    }
+    else if (bson) {
+        const auto& bson_data = bson->to_string();
+        auto str = new realm_string_t{bson_data.c_str(), bson_data.size()};
+        callback(data, str, nullptr);
+    }
+    delete_data(data);
```

Comment:
We should avoid imperatively deleting pointers like that. My suggestion is to declare a `UserdataPtr userdata(data, delete_data)` variable in the beginning of the function (where `UserdataPtr` is an alias for `std::unique_ptr<realm_userdata_t, realm_free_userdata_func_t>`). The `unique_ptr` destructor will fire and delete the userdata when the function unrolls even if any part of the following code throws an exception. Deleting at the end of the function will not be executed if the function prematurely unrolls.